src/System.IO.Compression/src/System/IO/Compression/DeflateManaged/DeflateManagedStream.cs

```diff
@@ -25,153 +27,89 @@ public partial class DeflateStream : Stream
                     break;
 
                 case CompressionMode.Compress:
-                    if (!stream.CanWrite)
-                    {
-                        throw new ArgumentException(SR.NotWriteableStream, "stream");
-                    }
-                    _deflater = CreateDeflater(null);
+                    InitializeDeflater(stream, leaveOpen, CompressionLevel.Optimal);
                     break;
 
                 default:
-                    throw new ArgumentException(SR.ArgumentOutOfRange_Enum, "mode");
+                    throw new ArgumentException(SR.ArgumentOutOfRange_Enum, nameof(mode));
             }
-
-            _stream = stream;
-            _mode = mode;
-            _leaveOpen = leaveOpen;
-            _buffer = new byte[DefaultBufferSize];
         }
 
         // Implies mode = Compress
-        public DeflateStream(Stream stream, CompressionLevel compressionLevel)
-
-            : this(stream, compressionLevel, false)
+        public DeflateManagedStream(Stream stream, CompressionLevel compressionLevel) : this(stream, compressionLevel, false)
         {
         }
 
         // Implies mode = Compress
-        public DeflateStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen)
+        public DeflateManagedStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen)
         {
             if (stream == null)
-                throw new ArgumentNullException("stream");
+                throw new ArgumentNullException(nameof(stream));
 
-            if (!stream.CanWrite)
-                throw new ArgumentException(SR.NotWriteableStream, "stream");
+            InitializeDeflater(stream, leaveOpen, compressionLevel);
+        }
 
-            // Checking of compressionLevel is passed down to the IDeflater implementation as it
-            // is a pugable component that completely encapsulates the meaning of compressionLevel.
 
-            Contract.EndContractBlock();
+        /// <summary>
+        /// Sets up this DeflateManagedStream to be used for Zlib Inflation/Decompression
+        /// </summary>
+        internal void InitializeInflater(Stream stream, bool leaveOpen, IFileFormatReader reader = null)
+        {
+            Debug.Assert(stream != null);
+            if (!stream.CanRead)
+                throw new ArgumentException(SR.NotSupported_UnreadableStream, nameof(stream));
+
+            _inflater = new InflaterManaged(reader);
 
             _stream = stream;
-            _mode = CompressionMode.Compress;
+            _mode = CompressionMode.Decompress;
             _leaveOpen = leaveOpen;
-
-            _deflater = CreateDeflater(compressionLevel);
-
             _buffer = new byte[DefaultBufferSize];
         }
 
-        private static IDeflater CreateDeflater(CompressionLevel? compressionLevel)
+        /// <summary>
+        /// Sets up this DeflateManagedStream to be used for Zlib Deflation/Compression
```

Comment:
I built this DeflateStream off of the existing one so they'd be minimally different if we decided to merge them later. I remove this comment in a later commit.